pkg/sysregistriesv2/system_registries_v2.go

```diff
@@ -338,55 +338,104 @@ func (config *V2RegistriesConf) postProcess() error {
-		return userRegistriesFilePath
-	}
-
-	if ctx != nil && ctx.RootForImplicitAbsolutePaths != "" {
-		return filepath.Join(ctx.RootForImplicitAbsolutePaths, systemRegistriesConfPath)
-	}
-
-	return systemRegistriesConfPath
+	wrapper := newConfigWrapper(ctx)
+	return wrapper.configPath
 }
 
 // ConfigDirPath returns the path to the system-wide directory for drop-in
 // registry configuration files.
+// Deprecated: This API only has support one directory from system-wide or per-user.
+// To get the support for file and directory override, we added the new
+// ConfigurationSourceDescription API. The new API should be used and this API is kept to
+// maintain backward compatibility.
 func ConfigDirPath(ctx *types.SystemContext) string {
-	if ctx != nil && ctx.SystemRegistriesConfDirPath != "" {
-		return ctx.SystemRegistriesConfDirPath
-	}
-
-	userRegistriesDirPath := filepath.Join(homedir.Get(), userRegistriesDir)
-	if _, err := os.Stat(userRegistriesDirPath); err == nil {
-		return userRegistriesDirPath
-	}
-
-	if ctx != nil && ctx.RootForImplicitAbsolutePaths != "" {
-		return filepath.Join(ctx.RootForImplicitAbsolutePaths, systemRegistriesConfDirPath)
+	wrapper := newConfigWrapper(ctx)
+	if wrapper.userConfigDirPath != "" {
+		return wrapper.userConfigDirPath
 	}
-
-	return systemRegistriesConfDirPath
+	return wrapper.configDirPath
 }
 
 // configWrapper is used to store the paths from ConfigPath and ConfigDirPath
 // and acts as a key to the internal cache.
 type configWrapper struct {
-	configPath    string
-	configDirPath string
+	configPath        string
+	configDirPath     string
+	userConfigDirPath string
 }
 
 // newConfigWrapper returns a configWrapper for the specified SystemContext.
 func newConfigWrapper(ctx *types.SystemContext) configWrapper {
-	return configWrapper{
-		configPath:    ConfigPath(ctx),
-		configDirPath: ConfigDirPath(ctx),
+	var wrapper configWrapper
+	userRegistriesFilePath := filepath.Join(homedir.Get(), userRegistriesFile)
+	userRegistriesDirPath := filepath.Join(homedir.Get(), userRegistriesDir)
+	if ctx != nil && ctx.SystemRegistriesConfPath != "" {
+		wrapper.configPath = ctx.SystemRegistriesConfPath
+		if _, err := os.Stat(userRegistriesDirPath); err == nil {
+			wrapper.userConfigDirPath = userRegistriesDirPath
+		}
+		if ctx.SystemRegistriesConfDirPath != "" {
+			wrapper.configDirPath = ctx.SystemRegistriesConfDirPath
+		} else if ctx.RootForImplicitAbsolutePaths != "" {
+			wrapper.configDirPath = filepath.Join(ctx.RootForImplicitAbsolutePaths, systemRegistriesConfDirPath)
+		} else {
+			wrapper.configDirPath = systemRegistriesConfDirPath
+		}
+		return wrapper
 	}
+	if _, err := os.Stat(userRegistriesFilePath); err == nil {
+		wrapper.configPath = userRegistriesFilePath
+		if _, err := os.Stat(userRegistriesDirPath); err == nil {
+			wrapper.userConfigDirPath = userRegistriesDirPath
+		}
+		return wrapper
+	}
+	if ctx != nil && ctx.RootForImplicitAbsolutePaths != "" {
```

Comment:
This is very similar to the `RootForImplicitAbsolutePaths == ""` variant, please combine them and only make the `wrapper.configPath` determination conditional.

It seems to me that we really want to make a “shall we use the system configuration file or `userRegistriesFilePath`” decision first, treat `userRegistriesFilePath` specially, and in all the other cases share the code for determining directories.

But that’s also up to what others think the behavior should be in the various cases.